byterun/shared_heap.c

```diff
@@ -188,25 +186,39 @@ static void pool_release(struct caml_heap_state* local, pool* pool, sizeclass sz
   Assert(pool->sz == sz);
   local->stats.pool_words -= POOL_WSIZE;
   local->stats.pool_frag_words -= POOL_HEADER_WSIZE + wastage_sizeclass[sz];
-  if (local->num_free_pools < MAX_LOCAL_FREE_POOLS) {
-    /* FIXME: in the local cache, other domains can't get it */
-    local->num_free_pools++;
-    pool->next = local->free_pools;
-    local->free_pools = pool;
-  } else {
-    /* FIXME: give free pools back to the OS */
-    caml_plat_lock(&pool_freelist.lock);
-    pool->next = pool_freelist.free;
-    pool_freelist.free = pool;
-    caml_plat_unlock(&pool_freelist.lock);
+  /* FIXME: give free pools back to the OS */
+  caml_plat_lock(&pool_freelist.lock);
+  pool->next = pool_freelist.free;
+  pool_freelist.free = pool;
+  caml_plat_unlock(&pool_freelist.lock);
+}
+
+static void calc_pool_stats(pool* a, sizeclass sz, struct heap_stats* s) {
+  value* p = (value*)((char*)a + POOL_HEADER_SZ);
+  value* end = (value*)a + POOL_WSIZE;
+  mlsize_t wh = wsize_sizeclass[sz];
+  s->pool_frag_words += Wsize_bsize(POOL_HEADER_SZ);
+
+  while (p + wh <= end) {
+    header_t hd = (header_t)*p;
+    if (hd) {
+      s->pool_live_words += Whsize_hd(hd);
+      s->pool_frag_words += wh - Whsize_hd(hd);
+      s->pool_live_blocks++;
+    }
+
+    p += wh;
   }
+  CAMLassert(end - p == wastage_sizeclass[sz]);
```

Comment:
This PR is introducing CAMLassert, but elsewhere in the file it is Assert, is this intentional?

Reply:
Looks like they're the same thing, will change back.